rust_src/src/keymap.rs

```diff
@@ -244,6 +245,70 @@ pub fn set_keymap_parent(keymap: LispObject, parent: LispObject) -> LispObject {
     parent
 }
 
+/// Return a copy of the keymap KEYMAP.
+///
+/// Note that this is almost never needed.  If you want a keymap that's like
+/// another yet with a few changes, you should use map inheritance rather
+/// than copying.  I.e. something like:
+///
+///     (let ((map (make-sparse-keymap)))
+///       (set-keymap-parent map <theirmap>)
+///       (define-key map ...)
+///       ...)
+///
+/// After performing `copy-keymap', the copy starts out with the same definitions
+/// of KEYMAP, but changing either the copy or KEYMAP does not affect the other.
+/// Any key definitions that are subkeymaps are recursively copied.
+/// However, a key definition which is a symbol whose definition is a keymap
+/// is not copied.
+#[lisp_fn]
+pub fn copy_keymap(mut keymap: LispObject) -> LispObject {
+    keymap = LispObject::from_raw(get_keymap(keymap.to_raw(), true, false));
+    let copy = list!(LispObject::from_raw(Qkeymap));
+    let mut tail = copy;
+    keymap = keymap.as_cons_or_error().cdr();
+
+    while let Some(keymap_cons) = keymap.as_cons() {
+        if !keymap_cons.car().eq_raw(Qkeymap) {
+            break;
+        }
+
+        let mut elt = LispObject::from_raw(unsafe { Fcopy_sequence(keymap_cons.car().to_raw()) });
+        if elt.is_char_table() {
+            unsafe { map_char_table(copy_keymap_1, Qnil, elt.to_raw(), elt.to_raw()) };
+        } else if elt.is_vector() {
+            if let Some(elt_pseudovec) = elt.as_vectorlike() {
+                let mut elt_vec = unsafe { elt.as_vector_unchecked() };
+                for i in 0..elt_pseudovec.pseudovector_size() {
+                    let copied_item =
+                        LispObject::from_raw(unsafe { copy_keymap_item(aref(elt, i).to_raw()) });
+                    elt_vec.set(i as usize, copied_item);
+                }
+            }
+        } else if let Some(elt_cons) = elt.as_cons() {
+            if elt.eq_raw(Qkeymap) {
+                elt = copy_keymap(elt); // This is a sub keymap.
+            } else {
+                let (elt_car, elt_cdr) = elt_cons.as_tuple();
+                elt = LispObject::from_raw(unsafe {
+                    Fcons(elt_car.to_raw(), copy_keymap_item(elt_cdr.to_raw()))
+                });
+            }
+        }
+
```

Comment:
This `if` implies `tail` could safely not be a cons cell. Which is not the case. The C code uses macros starting with 'X' to let you know they are not doing any type checking. Their intent is faster processing. This means one strategy would be to do this `as_cons().unwrap()`. Perhaps we could be a little smarter and use `unwrap_or_else(|| Error message)`??

Reply:
That makes sense. I assumed that the goal was to ensure safety all the time, but I understand the need for moving type checking to the side for faster processing.

How about using `as_cons_or_error()` instead? I've used it before and it seems to be doing what we want.

Reply:
Also, how does this look?

```rust
        } else if elt.is_cons() {
            let elt_cons = elt.as_cons_or_error();
            if elt.eq_raw(Qkeymap) {
                elt = copy_keymap(elt); // This is a sub keymap.
            } else {
                let (elt_car, elt_cdr) = elt_cons.as_tuple();
                elt = LispObject::from_raw(unsafe {
                    Fcons(elt_car.to_raw(), copy_keymap_item(elt_cdr.to_raw()))
                });
            }
        }

        let tail_cons = tail.as_cons_or_error();
        tail_cons.set_cdr(list!(elt));
        tail = tail_cons.cdr();
        keymap = keymap_cons.cdr();
```

Reply:
It is fine for now. We may need to revisit this once we are at a point where performance matters.